compiler/rustc_target/src/spec/maybe_lazy.rs

```diff
@@ -0,0 +1,137 @@
+//! A custom LazyLock+Cow suitable for holding borrowed, owned or lazy data.
+
+use std::borrow::{Borrow, Cow};
+use std::fmt::{Debug, Display};
+use std::ops::Deref;
+use std::sync::LazyLock;
+
+enum MaybeLazyInner<T: 'static + ToOwned + ?Sized> {
+    Lazy(LazyLock<<T as ToOwned>::Owned>),
+    Cow(Cow<'static, T>),
+}
+
+/// A custom LazyLock+Cow suitable for holding borrowed, owned or lazy data.
+///
+/// Technically this structure has 3 states: borrowed, owned and lazy
+/// They can all be constructed from the [`MaybeLazy::borrowed`], [`MaybeLazy::owned`] and
+/// [`MaybeLazy::lazy`] methods.
+#[repr(transparent)]
+pub struct MaybeLazy<T: 'static + ToOwned + ?Sized> {
+    // Inner state.
+    //
+    // Not to be inlined since we may want in the future to
+    // make this struct usable to statics and we might need to
+    // workaround const-eval limitation (particulary around drop).
+    inner: MaybeLazyInner<T>,
+}
+
+impl<T: 'static + ?Sized + ToOwned> MaybeLazy<T> {
+    /// Create a [`MaybeLazy`] from an borrowed `T`.
+    #[inline]
+    pub const fn borrowed(a: &'static T) -> Self {
+        Self { inner: MaybeLazyInner::Cow(Cow::Borrowed(a)) }
+    }
+
+    /// Create a [`MaybeLazy`] from an borrowed `T`.
+    #[inline]
+    pub const fn owned(a: <T as ToOwned>::Owned) -> Self {
+        Self { inner: MaybeLazyInner::Cow(Cow::Owned(a)) }
+    }
+
+    /// Create a [`MaybeLazy`] that is lazy by taking a function pointer.
+    ///
+    /// This function pointer cannot *ever* take a closure. User can potentially
+    /// workaround that by using closure-to-fnptr or `const` items.
+    #[inline]
+    pub const fn lazy(a: fn() -> <T as ToOwned>::Owned) -> Self {
+        Self { inner: MaybeLazyInner::Lazy(LazyLock::new(a)) }
+    }
+}
+
+impl<T: 'static + ?Sized + ToOwned> Clone for MaybeLazy<T>
```

Comment:
```suggestion
impl<T: 'static + ?Sized + ToOwned<Owned: Clone>> Clone for MaybeLazy<T>
```
In general, all `where` clauses below look convertible to simple bounds on parameters.